packages/utils/src/cmds/schema/index.js

```diff
@@ -0,0 +1,143 @@
+const fs = require('fs')
+const path = require('path')
+const { namespace } = require('@dhis2/cli-helpers-engine')
+const utils = require('../../support/utils')
+const request = require('request')
+const inquirer = require('inquirer')
+const { reporter } = require('@dhis2/cli-helpers-engine')
+
+const defaultOpts = {
+    schemasEndpoint: '/api/schemas.json',
+    infoEndpoint: '/api/system/info.json',
+}
+
+const defaultRequestOpts = {
+    headers: {
+        'x-requested-with': 'XMLHttpRequest',
+    },
+    json: true,
+}
+
+const prompt = inquirer.createPromptModule({ output: process.stderr })
+
+const schemaIdentifier = info => `${info.version}_${info.revision}`
+const schemaDiffIdentifier = (info1, info2) =>
+    `${schemaIdentifier(info1)}__${schemaIdentifier(info2)}`
+
+function asyncRequest(url, opts) {
+    return new Promise((resolve, reject) => {
+        request.get(url, opts, (err, res, body) => {
+            if (err || res.statusCode > 299) {
+                reporter.error('Request', url, 'failed to fetch')
+                err && reporter.dumpErr(err)
+                res &&
+                    reporter.error(
+                        res.statusCode,
+                        ':',
+                        res.statusMessage || (res.body && res.body.message)
+                    )
+                process.exit(1)
+            }
+            return resolve(body)
+        })
+    })
+}
+
+function resolveConfig(args) {
+    if (!args.utils && !args.utils.schema) {
+        return args
+    }
+    const config = args.utils.schema
+    return {
+        ...args,
+        ...config,
+        auth: args.auth || authFromConf(config),
+        authLeft: args.auth || authFromConf(config, config.leftServer),
+        authRight: args.auth || authFromConf(config, config.rightServer),
+    }
+}
+
+function authFromConf(conf = {}, serverConfig = {}) {
+    const auth = {
+        username: serverConfig.username || conf.username,
+        password: serverConfig.password || conf.password,
+    }
+    if (auth.username && auth.password) return auth
+    return true
+}
+
+async function getAuthHeader(url, { auth }) {
+    let { username, password } = auth
```

Comment:
This breaks if the tool is used without a `config` file, which we should support.  After changing `&&` to `||` above I get this error:

```
./packages/utils/bin/d2-utils schema --auth diff https://play.dhis2.org/dev http://localhost:8080
d2-utils schema diff <leftUrl> <rightUrl>

Generate a diff of schemas between DHIS2 versions

Positionals:
  leftUrl   the url to the running DHIS2 server, should have schemas available
            relative to this at /api/schemas.json
            Can also be a JSON-file with schemas. See schema fetch
                                                             [string] [required]
  rightUrl  the url to another running DHIS2 server, should have schemas
            available relative to this at /api/schemas.json
            Can also be a JSON-file with schemas. See "utils schema fetch"
                                                             [string] [required]

Options:
  --version             Show version number                            [boolean]
  --config              Path to JSON config file
  --force               By default each schema is cached, identified by the
                        version and revision. Use this flag to disable caching
                        and download the schemas.                      [boolean]
  --auth                Prompt credentials for each server.
                                                      [boolean] [default: false]
  --base-url, -b        BaseUrl to use for downloading schemas. If this is set
                        leftServer and rightServer should be relative to this
                        url, eg. /dev.                                  [string]
  --output, -o          Specify the location of the output. If used as a flag a
                        file relative to current working directory is generated
                        with the name
                        "LEFT-version_revision__RIGHT-version_revision.html".
                        If the location is a directory, the default filename is
                        used and output to location.                    [string]
  --format              Specify the format of the output
              [string] [choices: "html", "json", "console"] [default: "console"]
  --ignore-array-order  Sort all nested arrays in schemas. The order of nested
                        arrays are non-deterministic, which may clutter the
                        diff. Enabling this will prevent most internal array
                        moves, which are probably irrelevant anyway.
                                                      [boolean] [default: false]
  -h, --help            Show help                                      [boolean]

TypeError: Cannot destructure property `username` of 'undefined' or 'null'.
    at getAuthHeader (/Users/awm/dev/dhis2/cli/d2/packages/utils/src/cmds/schema/index.js:70:34)
    at schemasFromUrl (/Users/awm/dev/dhis2/cli/d2/packages/utils/src/cmds/schema/index.js:95:47)
    at getSchemas (/Users/awm/dev/dhis2/cli/d2/packages/utils/src/cmds/schema/diff/index.js:47:25)
    at Object.run [as handler] (/Users/awm/dev/dhis2/cli/d2/packages/utils/src/cmds/schema/diff/index.js:188:31)
    at Object.runCommand (/Users/awm/dev/dhis2/cli/d2/node_modules/yargs/lib/command.js:242:26)
    at Object.parseArgs [as _parseArgs] (/Users/awm/dev/dhis2/cli/d2/node_modules/yargs/yargs.js:1078:30)
    at Object.runCommand (/Users/awm/dev/dhis2/cli/d2/node_modules/yargs/lib/command.js:200:96)
    at Object.parseArgs [as _parseArgs] (/Users/awm/dev/dhis2/cli/d2/node_modules/yargs/yargs.js:1078:30)
    at Object.parse (/Users/awm/dev/dhis2/cli/d2/node_modules/yargs/yargs.js:542:19)
    at module.exports (/Users/awm/dev/dhis2/cli/d2/node_modules/@dhis2/cli-helpers-engine/lib/makeEntryPoint.js:43:11)
```

This comes originally from `diff/index.js:190`.  I think some more testing of un-configured schema diffing might be in order!

Reply:
Sorry, I forgot to fallback to prompt if no auth was given. Should be fixed now.